components/tidb_query/src/rpn_expr/impl_compare_in.rs

```diff
@@ -0,0 +1,366 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::collections::HashSet;
+use std::hash::Hash;
+
+use codec::prelude::NumberDecoder;
+use tidb_query_codegen::rpn_fn;
+use tidb_query_datatype::EvalType;
+use tipb::{Expr, ExprType};
+
+use crate::codec::data_type::*;
+use crate::codec::mysql::{Decimal, MAX_FSP};
+use crate::{Error, Result};
+
+pub trait Extract: std::marker::Sized {
+    fn extract(expr_tp: ExprType, val: Vec<u8>) -> Result<Option<Self>>;
+}
+
+#[inline]
+fn type_error(eval_type: EvalType, expr_type: ExprType) -> Error {
+    return other_err!(
+        "Unexpected ExprType {:?} and EvalType {:?}",
+        expr_type,
+        eval_type
+    );
+}
+
+impl Extract for Int {
+    #[inline]
+    fn extract(expr_tp: ExprType, val: Vec<u8>) -> Result<Option<Self>> {
+        if expr_tp == ExprType::Int64 {
+            let value = val
+                .as_slice()
+                .read_i64()
+                .map_err(|_| other_err!("Unable to decode int64 from the request"))?;
+            Ok(Some(value))
+        } else if expr_tp == ExprType::Uint64 {
+            let value = val
+                .as_slice()
+                .read_u64()
+                .map_err(|_| other_err!("Unable to decode uint64 from the request"))?;
+            Ok(Some(value as i64))
+        } else {
+            Err(type_error(Int::EVAL_TYPE, expr_tp))
+        }
+    }
+}
+
+impl Extract for Real {
+    #[inline]
+    fn extract(expr_tp: ExprType, val: Vec<u8>) -> Result<Option<Self>> {
+        if expr_tp != ExprType::MysqlDecimal {
+            return Err(type_error(Real::EVAL_TYPE, expr_tp));
+        }
+        let value = val
+            .as_slice()
+            .read_f64()
+            .map_err(|_| other_err!("Unable to decode float from the request"))?;
+        Ok(Real::new(value).ok())
```

Comment:
I think we can return Error instead of None in this case, so that you won't need `has_null = true;` in L185.